File: 5-longest-palindromic-substring/longest-palindromic-substring.py
from typing import Tuple

class Solution:
    def longestPalindrome(self, s: str) -> str:
        def longest_odd_palindrome(start_index: int) -> Tuple[int, int]:
            i, j = start_index - 1, start_index + 1
            if i < 0 or j > len(s):
                return start_index, start_index
            
            while i >= 0 and j < len(s) and s[i] == s[j]:
                i -= 1
                j += 1
            
            return i + 1, j - 1
        
        def longest_even_palindrome(start_index: int) -> Tuple[int, int]:
            i, j = start_index - 1, start_index
            if i < 0:
                return -1, -1
            while i >= 0 and j < len(s) and s[i] == s[j]:
                i -= 1
                j += 1
            
            return i + 1, j - 1
        
        longest_start = 0
        longest_end = 0

        for i in range(len(s)):
            odd_start, odd_end = longest_odd_palindrome(i)
            if odd_end - odd_start > longest_end - longest_start:
                longest_start, longest_end = odd_start, odd_end
        
            even_start, even_end = longest_even_palindrome(i)
            if even_end - even_start > longest_end - longest_start:
                longest_start, longest_end = even_start, even_end
        
        return s[longest_start: longest_end + 1]

File: 5-longest-palindromic-substring/longest-palindromic-substring.go
// get the longest odd len palindrome containing s[i] at middle
func getOddLenPalindrome(s string, index int) (int, int) {
    left, right := index - 1, index + 1
    if left < 0 || right >= len(s) {
        return index, index
    }

    for left >= 0 && right < len(s) && s[left] == s[right]{
        left--
        right++
    }

    return left + 1, right - 1
}


// get the longest even len palindrome containing s[i] at middle
func getEvenLenPalindrome(s string, index int) (int, int) {
    left, right := index, index + 1
    if right >= len(s) {
        return index, index
    }

    for left >= 0 && right < len(s) && s[left] == s[right]{
        left--
        right++
    }

    // fmt.Println(index, left + 1, right - 1)

    return left + 1, right - 1
}


func longestPalindrome(s string) string {
    longestStart, longestEnd := 0, 0
    for i := range len(s){
        oddStart, oddEnd := getOddLenPalindrome(s, i)
        evenStart, evenEnd := getEvenLenPalindrome(s, i)
        // fmt.Println("i:", i)
        // fmt.Println("Even:", evenStart, evenEnd)
        // fmt.Println("Odd:", oddStart, oddEnd)

        if oddEnd - oddStart > longestEnd - longestStart{
            longestStart = oddStart
            longestEnd = oddEnd
        }
        if evenEnd - evenStart > longestEnd - longestStart{
            longestStart = evenStart
            longestEnd = evenEnd
        }

        // fmt.Println(longestStart, longestEnd)

    }

    return s[longestStart:longestEnd + 1]
}